File: apps/web/src/lib/issue-estimate.ts
// EXP-630: story points, rendered on the TEAM's scale (`teams.estimation_type`,
// contract `issueEstimation`). The stored value is always a point number, so
// a team can switch scales without touching a single issue; `none` hides the
// chip everywhere. T-shirt sizes are the fibonacci points worn as XS…XL,
// exactly how Linear stores them.
import {
  ISSUE_ESTIMATE_TSHIRT_LABELS,
  ISSUE_ESTIMATION_SCALES,
  type IssueEstimation,
} from "@/lib/domain"

export type EstimationScale = Exclude<IssueEstimation, `none`>

/** The settings picker rows, in Linear's order and wording. */
export const ESTIMATION_TYPE_OPTIONS: { value: IssueEstimation; label: string; hint: string }[] = [
  { value: `none`, label: `Not in use`, hint: `` },
  { value: `exponential`, label: `Exponential`, hint: `1, 2, 4, 8, 16 points` },
  { value: `fibonacci`, label: `Fibonacci`, hint: `1, 2, 3, 5, 8 points` },
  { value: `linear`, label: `Linear`, hint: `1, 2, 3, 4, 5 points` },
  { value: `tshirt`, label: `T-shirt`, hint: `XS, S, M, L, XL` },
]

export const NO_ESTIMATE = `No estimate`

export function estimationScale(type: IssueEstimation | string): readonly number[] {
  return type in ISSUE_ESTIMATION_SCALES
    ? ISSUE_ESTIMATION_SCALES[type as EstimationScale]
    : []
}

function tshirtLabel(value: number): string | null {
  const index = ISSUE_ESTIMATION_SCALES.tshirt.indexOf(value)
  return index === -1 ? null : ISSUE_ESTIMATE_TSHIRT_LABELS[index]!
}

/** "No estimate", "M", "1 point", "5 points". */
export function estimateLabel(
  value: number | null | undefined,
  type: IssueEstimation | string = `fibonacci`
): string {
  if (value === null || value === undefined) return NO_ESTIMATE
  if (type === `tshirt`) {
    const size = tshirtLabel(value)
    if (size) return size
  }
  return value === 1 ? `1 point` : `${value} points`
}

/** The chip form: "M" on the t-shirt scale, "5 pt" elsewhere. */
export function estimateShortLabel(value: number, type: IssueEstimation | string = `fibonacci`): string {
  if (type === `tshirt`) {
    const size = tshirtLabel(value)
    if (size) return size
  }
  return `${value} pt`
}

/**
 * The values a picker offers: the scale's ladder plus the current value when
 * it sits off the ladder (an import from another scale, a value set before
 * the scale was switched), ascending — so the trigger always names a listed
 * option. The "No estimate" row is the caller's.
 */
export function estimatePickerValues(
  current: number | null | undefined,
  type: IssueEstimation | string
): number[] {
  const values = new Set<number>(estimationScale(type))
  if (typeof current === `number` && current >= 0) values.add(current)
  return [...values].sort((left, right) => left - right)
}

/**
 * Picker options: "No estimate" first, then `estimatePickerValues`. Values
 * are strings (the picker's currency); `""` = clear.
 */
export function estimatePickerOptions(
  current: number | null | undefined,
  type: IssueEstimation
): { value: string; label: string }[] {
  return [
    { value: ``, label: NO_ESTIMATE },
    ...estimatePickerValues(current, type).map((value) => ({
      value: String(value),
      label: estimateLabel(value, type),
    })),
  ]
}

/** The `estimate_changed` timeline phrase, on the team's scale. */
export function estimateEventPhrase(
  payload: Record<string, unknown> | null | undefined,
  type: IssueEstimation | string = `fibonacci`
): string {
  const raw = payload?.to
  const to =
    typeof raw === `number` ? raw : typeof raw === `string` && raw !== `` ? Number(raw) : null
  if (to === null || !Number.isFinite(to)) return `removed the estimate`
  return `set the estimate to ${estimateLabel(to, type)}`
}

/** The picker's string back to the wire value (`""` → null). */
export function parseEstimatePick(value: string | null | undefined): number | null {
  if (!value) return null
  const parsed = Number.parseInt(value, 10)
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : null
}

File: apps/android/app/src/main/java/com/exponential/app/domain/IssueEstimate.kt
package com.exponential.app.domain

import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

// EXP-630: story points, rendered on the TEAM's scale (`teams.estimation_type`,
// contract `issueEstimation`). The stored value is always a point number, so
// a team can switch scales without touching a single issue; `none` hides the
// chip everywhere. T-shirt sizes are the fibonacci points worn as XS…XL,
// exactly how Linear stores them. Mirrors web `lib/issue-estimate.ts`; locked
// ×4 by `domain-contract/fixtures/issue-estimate.json` (IssueEstimateTest).

/** The ladders per scale, in the contract's order (`none` has none). */
val ISSUE_ESTIMATION_SCALES: Map<String, List<Int>> = linkedMapOf(
    DomainContract.issueEstimationExponential to listOf(1, 2, 4, 8, 16),
    DomainContract.issueEstimationFibonacci to listOf(1, 2, 3, 5, 8),
    DomainContract.issueEstimationLinear to listOf(1, 2, 3, 4, 5),
    DomainContract.issueEstimationTshirt to listOf(1, 2, 3, 5, 8),
)

/** The t-shirt ladder's sizes, index-aligned with its points. */
val ISSUE_ESTIMATE_TSHIRT_LABELS: List<String> = listOf("XS", "S", "M", "L", "XL")

const val NO_ESTIMATE = "No estimate"

/** The ladder of one scale; an unknown scale (or `none`) offers nothing. */
fun estimationScale(type: String): List<Int> = ISSUE_ESTIMATION_SCALES[type] ?: emptyList()

private fun tshirtLabel(value: Int): String? {
    val index = ISSUE_ESTIMATION_SCALES.getValue(DomainContract.issueEstimationTshirt).indexOf(value)
    return if (index == -1) null else ISSUE_ESTIMATE_TSHIRT_LABELS[index]
}

/** "No estimate", "M", "1 point", "5 points". */
fun estimateLabel(value: Int?, scale: String = DomainContract.issueEstimationFibonacci): String {
    if (value == null) return NO_ESTIMATE
    if (scale == DomainContract.issueEstimationTshirt) {
        tshirtLabel(value)?.let { return it }
    }
    return if (value == 1) "1 point" else "$value points"
}

/** The chip form: "M" on the t-shirt scale, "5 pt" elsewhere. */
fun estimateShortLabel(value: Int, scale: String = DomainContract.issueEstimationFibonacci): String {
    if (scale == DomainContract.issueEstimationTshirt) {
        tshirtLabel(value)?.let { return it }
    }
    return "$value pt"
}

/**
 * The values a picker offers: the scale's ladder plus the current value when
 * it sits off the ladder (an import from another scale, a value set before
 * the scale was switched), ascending — so the trigger always names a listed
 * option. The "No estimate" row is the caller's.
 */
fun estimatePickerValues(current: Int?, scale: String): List<Int> {
    val values = estimationScale(scale).toMutableSet()
    if (current != null && current >= 0) values.add(current)
    return values.sorted()
}

/** The `estimate_changed` timeline phrase, on the team's scale. */
fun estimateEventPhrase(
    payload: JsonObject?,
    scale: String = DomainContract.issueEstimationFibonacci,
): String {
    val to = estimatePayloadValue(payload?.get("to")) ?: return "removed the estimate"
    return "set the estimate to ${estimateLabel(to, scale)}"
}

// A payload side as a point value: a JSON number, or a numeric string (an
// older writer's stringified payload); null/absent/blank/NaN = no estimate.
private fun estimatePayloadValue(raw: kotlinx.serialization.json.JsonElement?): Int? {
    if (raw == null || raw is JsonNull) return null
    val content = (raw as? JsonPrimitive)?.contentOrNull?.takeIf { it.isNotBlank() } ?: return null
    content.toIntOrNull()?.let { return it }
    return content.toDoubleOrNull()?.takeIf { it.isFinite() }?.toInt()
}
